cutlery/_bbpe.pyx

```diff
@@ -0,0 +1,137 @@
+# cython: infer_types
+
+from typing import Dict, List, Optional, Tuple
+from cython.operator cimport dereference as deref
+from functools import lru_cache
+import json
+from libcpp.memory cimport make_shared, shared_ptr
+from pathlib import Path
+import regex
+
+
+# We need the same splitting pattern and byte -> string remapping as GPT-2,
+# so these are taken directly from:
+#
+# https://github.com/openai/gpt-2/blob/master/src/encoder.py
+#
+# See the LICENSE file for licensing information.
+
+
+SPLIT_PATTERN = r"""'s|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+"""
+
+
+@lru_cache()
+def bytes_to_unicode() -> Dict[int, str]:
+    """
+    Returns list of utf-8 byte and a corresponding list of unicode strings.
+    The reversible bpe codes work on unicode strings.
+    This means you need a large # of unicode characters in your vocab if you want to avoid UNKs.
+    When you're at something like a 10B token dataset you end up needing around 5K for decent coverage.
+    This is a signficant percentage of your normal, say, 32K bpe vocab.
+    To avoid that, we want lookup tables between utf-8 bytes and unicode strings.
+    And avoids mapping to whitespace/control characters the bpe code barfs on.
+    """
+    bs = (
+        list(range(ord("!"), ord("~") + 1))
+        + list(range(ord("¡"), ord("¬") + 1))
+        + list(range(ord("®"), ord("ÿ") + 1))
+    )
+    cs = bs[:]
+    n = 0
+    for b in range(2**8):
+        if b not in bs:
+            bs.append(b)
+            cs.append(2**8 + n)
+            n += 1
+    cs = [chr(n).encode("utf-8") for n in cs]
+    return dict(zip(bs, cs))
+
+cdef class ByteBPEProcessor:
+    cdef shared_ptr[Merges] _merges
+    cdef dict _byte_decoder
+    cdef dict _byte_encoder
+    cdef object _split_pattern
+    cdef dict _vocab
+
+    def __init__(self, vocab: Dict[str, int], merges: List[Tuple[str, str]]):
+        self._byte_encoder = bytes_to_unicode()
+        self._byte_decoder = {v: k for k, v in self._byte_encoder.items()}
+        self._split_pattern = regex.compile(SPLIT_PATTERN)
+        self._vocab = vocab
+        cdef vector[pair[string, string]] c_merges
+        for p1, p2 in merges:
+            c_merges.push_back(pair[string, string](p1.encode('utf-8'), p2.encode('utf-8')))
+        self._merges = make_shared[Merges](c_merges)
+
+    @staticmethod
+    def load_from_files(*, vocab: Path, merges: Path) -> ByteBPEProcessor:
+        """Construct a processor from the given vocabulary and merges files."""
+        with open(vocab, encoding="utf-8") as f:
+            vocab = json.load(f)
+        with open(merges, encoding="utf-8") as f:
+            version = f.readline()
+            if not version.startswith("#version: 0.2"):
+                raise ValueError(f"Only version 0.2 of the merges format is supported, was: {version.strip()}")
```

Comment:
Me neither, but all the merges files that I encountered start with this header. Some also have some trailing material (like it being trained by HF transformers). So I thought it's good to add a check just in case someone changes the format.